async-embedded/src/lib.rs

```diff
@@ -0,0 +1,102 @@
+pub(crate) unsafe fn signal_event_ready() {
+    asm::sev();
+}
+
+#[cfg(target_arch = "arm")]
+#[inline]
+/// Wait for an interrupt or until notified by other hart via `signal_task_ready`
+/// This particular implementation does nothing
+pub(crate) unsafe fn wait_for_event() {
+    asm::wfe();
+}
+
+#[cfg(any(target_arch = "riscv32", target_arch = "riscv64"))]
+/// This keeps dropping into the debugger and never returns
+pub fn abort() -> ! {
+    loop {
+        unsafe { riscv::asm::ebreak() }
+    }
+}
+
+#[cfg(all(any(target_arch = "riscv32", target_arch = "riscv64"), feature = "riscv-wait-nop"))]
+#[inline]
+/// Prevent next `wait_for_interrupt` from sleeping, wake up other harts if needed.
+/// This particular implementation does nothing, since `wait_for_interrupt` never sleeps
+pub(crate) unsafe fn signal_event_ready() {}
+
+#[cfg(all(any(target_arch = "riscv32", target_arch = "riscv64"), feature = "riscv-wait-nop"))]
+#[inline]
+/// Wait for an interrupt or until notified by other hart via `signal_task_ready`
+/// This particular implementation does nothing
+pub(crate) unsafe fn wait_for_event() {}
+
+#[cfg(all(any(target_arch = "riscv32", target_arch = "riscv64"), feature = "riscv-wait-extern"))]
+extern "C" {
+    /// Prevent next `wait_for_interrupt` from sleeping, wake up other harts if needed.
+    /// User is expected to provide an actual implementation, like the one shown below.
+    ///
+    /// #[no_mangle]
+    /// pub extern "C" fn signal_event_ready() {
+    ///     unimplemented!();
+    /// }
+    pub(crate) fn signal_event_ready();
+
+    /// Wait for an interrupt or until notified by other hart via `signal_task_ready`
+    /// User is expected to provide an actual implementation, like the one shown below.
+    ///
+    /// #[no_mangle]
+    /// pub extern "C" fn wait_for_event() {
+    ///     unimplemented!();
+    /// }
+    pub(crate) fn wait_for_event();
+}
+
+#[cfg(all(any(target_arch = "riscv32", target_arch = "riscv64"), feature = "riscv-wait-wfi-single-hart"))]
+static mut TASK_READY: bool = false;
+
+#[cfg(all(any(target_arch = "riscv32", target_arch = "riscv64"), feature = "riscv-wait-wfi-single-hart"))]
+#[inline]
+/// Prevent next `wait_for_interrupt` from sleeping, wake up other harts if needed.
+/// This particular implementation prevents `wait_for_interrupt` from sleeping by setting
+/// a global mutable flag
+pub(crate) unsafe fn signal_event_ready() {
+    TASK_READY = true;
+}
+
+#[cfg(all(any(target_arch = "riscv32", target_arch = "riscv64"), feature = "riscv-wait-wfi-single-hart"))]
+#[inline]
+/// Wait for an interrupt or until notified by other hart via `signal_task_ready`
+/// This particular implementation decides whether to sleep or not by checking
+/// a global mutable flag that's set by `signal_task_ready`
+pub(crate) unsafe fn wait_for_event() {
+    if !TASK_READY {
+        riscv::asm::wfi();
+        TASK_READY = false;
+    }
```

Comment:
```suggestion
    }
    TASK_READY = false;
```

Oops, I found a mistake

Reply:
The "riscv-wait-wfi-single-hart" impl already relies on the fact that you only `wait_for_event()/signal_event_ready()` from thread context. So, I don't see how `TASK_READY` could be set to `true` during `wait_for_event()`.

Reply:
Currently, `TASK_READY = true` sticks if you do not have `wait_for_event()` running "at the same time". Once set, there is not way to reset it back.